rust-src/bulletproofs/src/set_non_membership_proof.rs

```diff
@@ -0,0 +1,674 @@
+    let mut set_vec = the_set.to_vec();
+    pad_vector_to_power_of_two(&mut set_vec);
+    let n = set_vec.len();
+    if gens.G_H.len() < n {
+        return Err(VerificationError::NotEnoughGenerators);
+    }
+    // Select generators for vector commitments
+    let (G, H): (Vec<_>, Vec<_>) = gens.G_H.iter().take(n).cloned().unzip();
+
+    // Domain separation
+    transcript.add_bytes(b"SetNonMembershipProof");
+    // append commitment V to transcript
+    transcript.append_message(b"V", &V.0);
+    transcript.append_message(b"theSet", &set_vec);
+
+    // define the commitments A,S
+    let A = proof.A;
+    let S = proof.S;
+    // append commitments A and S to transcript
+    transcript.append_message(b"A", &A);
+    transcript.append_message(b"S", &S);
+
+    // get challenges y,z from transcript
+    let y: C::Scalar = transcript.challenge_scalar::<C, _>(b"y");
+    let z: C::Scalar = transcript.challenge_scalar::<C, _>(b"z");
+
+    // define the commitments T1, T2
+    let T_1 = proof.T_1;
+    let T_2 = proof.T_2;
+    // append T1, T2 commitments to transcript
+    transcript.append_message(b"T1", &T_1);
+    transcript.append_message(b"T2", &T_2);
+
+    // get challenge x (evaluation point) from transcript
+    let x: C::Scalar = transcript.challenge_scalar::<C, _>(b"x");
+
+    // define polynomial evaluation value
+    let tx = proof.tx;
+    // define blinding factors for tx and IP proof
+    let tx_tilde = proof.tx_tilde;
+    let e_tilde = proof.e_tilde;
+    // append tx, tx_tilde, e_tilde to transcript
+    transcript.append_message(b"tx", &tx);
+    transcript.append_message(b"tx_tilde", &tx_tilde);
+    transcript.append_message(b"e_tilde", &e_tilde);
+
+    // get challenge w from transcript
+    let w: C::Scalar = transcript.challenge_scalar::<C, _>(b"w");
+
+    // compute delta(y,z) = <1, y^n>  - z<s, y^n>
+    let yn = z_vec(y, 0, n);
+    let one_vec = vec![C::Scalar::one(); n];
+    let ip_one_yn = inner_product(&one_vec, &yn);
+    let mut zip_s_yn = inner_product(&set_vec, &yn);
+    zip_s_yn.mul_assign(&z);
+    let mut delta_yz = ip_one_yn;
+    delta_yz.sub_assign(&zip_s_yn);
+
+    // Part 2: Verify consistency of t_0
+    // i.e., check 0 = V^(z <1, yn>) * g^(delta_yz - tx) * T_1^x * T_2^x^2 *
+    // h^(-tx_tilde)
+    let mut zip_one_yn = ip_one_yn;
+    zip_one_yn.mul_assign(&z);
+    let mut delta_minus_tx = delta_yz;
+    delta_minus_tx.sub_assign(&tx);
+    let mut x2 = x; // x^2
+    x2.mul_assign(&x);
+    let mut minus_tx_tilde = tx_tilde;
+    minus_tx_tilde.negate();
+
+    let t0_check_base_points = vec![V.0, v_keys.g, T_1, T_2, v_keys.h];
+    let t0_check_exponents = vec![zip_one_yn, delta_minus_tx, x, x2, minus_tx_tilde];
+
+    let rhs = multiexp(&t0_check_base_points, &t0_check_exponents);
+    if !rhs.is_zero_point() {
+        return Err(VerificationError::InconsistentT0);
+    }
+
+    // Part 3: Verify inner product
+    // First compute helper variables g_hat, h_prime, and P_prime
```

Comment:
Any reason why we call it g_hat here and Q in the prover side?

Reply:
Legacy. It used to be Q everywhere for the range proof because it comes from Dalek. I always tried to use bluepaper notation instead and sometimes added comments saying that this corresponds to that in bluepaper.